File: lib/teien/actors/box_actor.rb
module Teien
  class BoxActor < Teien::Actor
    
    def initialize(name, vec = Teien::Vector3D.new(1, 1, 1), physics_info = nil, ext_info = nil)
      super(name, physics_info, ext_info)              
      @type = 'Box'
      @vec = vec
    end     
            
    def setup(physics)
      # create a physics object.
      @col_obj = Bullet::BtBoxShape.new(Bullet::BtVector3.new(@vec.x / 2.0, @vec.y / 2.0, @vec.z / 2.0))
      inertia = Bullet::BtVector3.new
      @col_obj.calculate_local_inertia(@physics_info[:mass], inertia)
      
      @rigid_body = Bullet::BtRigidBody.new(@physics_info[:mass], self, @col_obj, inertia)
      super(physics)      
      physics.add_rigid_body(@rigid_body)      
    end

    def set_angular_factor(vec)
      @rigid_body.set_angular_factor(Bullet::BtVector3.new(vec.x, vec.y, vec.z)) 
    end 
    
    def to_hash
      hash = super
      hash[:type] = @type      
      hash[:x] = @vec.x
      hash[:y] = @vec.y
      hash[:z] = @vec.z      
      hash
    end
    
    def from_hash(hash)
      super
      @vec.x = hash[:x]
      @vec.y = hash[:y]
      @vec.z = hash[:z]
    end
  end
end

File: public/teien/actors/box_actor.js
BoxActor = function(param){
    teien.Actor.call(this, param);
    this.type = param.type;
    this.x = param.x;
    this.y = param.y;
    this.z = param.z;    
};
BoxActor.prototype = Object.create(teien.Actor.prototype);

BoxActor.prototype.setup = function(physics){
    var col_obj = new Ammo.btBoxShape(new Ammo.btVector3(this.x / 2, this.y / 2, this.z / 2));
    var inertia = new Ammo.btVector3();
    col_obj.calculateLocalInertia(this.physics_info.mass, inertia);
    
    // Ammo does not support btMotionState handling.
    //this.rigidBody = new Ammo.btRigidBody(this.mass, this, col_obj, inertia);
    var ms = new Ammo.btDefaultMotionState(this.transform);
    this.rigidBody = new Ammo.btRigidBody(this.physics_info.mass, ms, col_obj, inertia);
    physics.addRigidBody(this.rigidBody);

    teien.Actor.prototype.setup.call(this, physics);    
};

BoxActor.prototype.fromHash = function(param){
    teien.Actor.prototype.fromHash.call(this, param);
    this.x = param.x;
    this.y = param.y;
    this.z = param.z;    
};

BoxActor.prototype.update = function(params){
    
};
